### PR TITLE
[dynamo] mutable local caching to make dynamo faster at tracing mutation

### PR DESCRIPTION
**Pull request by mlazos**
Make mutation faster to speed up tracing optimizers, helps with https://github.com/pytorch/torchdynamo/issues/1803

`replace_all` no longer iterates over the entire variable tracker data structure  every time a mutation is performed

Each variable tracker internally keeps a set of contained mutable variable trackers, to provide a hint to `replace_all`. This is populated with a call to `apply` from `__post_init__` in the base `VariableTracker`

cc @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/variables/base.py [hunk 1]
```diff
@@ -84,21 +100,29 @@ def apply(
...
+            if skip_fn is None or not skip_fn(value):
+                updated_dict = dict(value.__dict__)
+                for key in updated_dict.keys():
+                    if key not in value._nonvar_fields:
+                        updated_dict[key] = cls.apply(
+                            fn, updated_dict[key], cache, skip_fn
+                        )
+                result = fn(value.clone(**updated_dict))
+            else:
+                result = fn(value)
```
Comment (jansel) [inline torch/_dynamo/variables/base.py:112]:
```suggestion
                result = value
```
?

**Review by jansel — CHANGES_REQUESTED** [review]
1) How much does this help compile times do you have some numbers?

2) I worry this change will be error-prone and hard to maintain.  `recursively_contains` is updated in lots of different places, and if someone changes the code it will be super easy to miss something and get some super tricky to fix bugs as a result.

I'd suggest instead modifying TensorVariable.apply() to be much less aggressive about copying, then use the skip_fn approach with a cache on the object.

File: torch/_dynamo/variables/base.py (hunk 1 above)
Comment (mlazos) [reply torch/_dynamo/variables/base.py:112]: Isn't it possible I'm going to replace value?

**Comment by mlazos** [comment]
> 1. How much does this help compile times do you have some numbers?
> 2. I worry this change will be error-prone and hard to maintain.  `recursively_contains` is updated in lots of different places, and if someone changes the code it will be super easy to miss something and get some super tricky to fix bugs as a result.
> 
> I'd suggest instead modifying TensorVariable.apply() to be much less aggressive about copying, then use the skip_fn approach with a cache on the object.

1. Numbers ([script](https://gist.github.com/mlazos/6547fcfcf30f8a9da9d077d7d572dfd7)):

|Num params |Master (s)| With cache (s)| 
|---|---|---|
| 100 | 145.005 |20.025 |
| 200|630.251|74.029|
| 300|1633.136|204.839|
| 400| 3215.020|386.457|

2. Yeah I sent you some messages in chat, for others to see: I generally tend to agree, this is complex. I think in order to fix the fragility I can change this so that there is an optional argument for `dicts` and `lists` call it `recursive_contains` where the default is `None`. If it's `None` we defer to the `VariableTracker` which will iterate over its own `__dict__` post-init to populate the internal `_recursive_contains` field on the object. Then this will allow the default behavior to be correct with no special handling, and then for lists and dicts we can override this behavior for `append` and `__setitem__` to make it more efficient (ie only adding the incremental mutables from the new items getting added to the container) by passing in updated `recursive_contains`

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**